Generate serialization for NSPresentationIntent

mscottapple opened the pull request:
#### 089245fbdcadc21c8ef020e0a1ea4dd297355b4e
<pre>
Generate serialization for NSPresentationIntent
<a href="https://bugs.webkit.org/show_bug.cgi?id=268779">https://bugs.webkit.org/show_bug.cgi?id=268779</a>
<a href="https://rdar.apple.com/109121874">rdar://109121874</a>

Reviewed by achristensen07 (Alex Christensen).

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::typeFromObject):
(IPC::shouldEnableStrictMode):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.h: Added.
(WebKit::CoreIPCPresentationIntent::CoreIPCPresentationIntent):
* Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.mm: Added.
(WebKit::CoreIPCPresentationIntent::CoreIPCPresentationIntent):
(WebKit::CoreIPCPresentationIntent::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCTypes.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/274269@main">https://commits.webkit.org/274269@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1dd1fd2476454a738622e7288d0cd79a7213c4f4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38509 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17444 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41051 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40815 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20226 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14789 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39082 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14690 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12767 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34360 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42327 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34971 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38584 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/13358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11042 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36788 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14958 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5014 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14430 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -859,9 +860,9 @@ static void destroyTempKeychain(SecKeychainRef keychainRef)
     contact.get().phoneNumber = phoneNumber.get();
     contact.get().postalAddress = address.get();
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    ALLOW_DEPRECATED_DECLARATIONS_BEGIN
```
This indentation should probably not be changed.

mscottapple replied:
Grrr! Xcode keeps doing that to me! I've changed that in every single PR where I've touched this file.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.h:
```diff
@@ -0,0 +1,76 @@
+    {
+    }
+
+    int64_t m_intentKind = 0;
```
initializer list syntax is a little more modern:
int64_t m_intentKind { 0 };

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.h:
```diff
@@ -0,0 +1,76 @@
+#if PLATFORM(COCOA)
+
+#include "ArgumentCodersCocoa.h"
+#include <Foundation/Foundation.h>
```
This can probably be replaced by OBJC_CLASS NSPresentationIntent; for the header, then include this in the .mm file.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.h:
```diff
@@ -0,0 +1,76 @@
+private:
+    friend struct IPC::ArgumentCoder<CoreIPCPresentationIntent, void>;
+
+    CoreIPCPresentationIntent(int64_t intentKind, int64_t identity, std::unique_ptr<CoreIPCPresentationIntent>&& parentIntent, int64_t column, Vector<int64_t>&& columnAlignments, int64_t columnCount, int64_t headerLevel, const String& languageHint, int64_t ordinal, int64_t row)
```
Do these all need to be signed?  It seems like something like columnCount can't be negative.

mscottapple replied:
Yeah, I know, but the `NSPresentationIntent` implementation uses `NSInteger` for all these, so I followed suit.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.mm:
```diff
@@ -0,0 +1,126 @@
+
+RetainPtr<id> CoreIPCPresentationIntent::toID() const
+{
+    id parent = (m_parentIntent ? m_parentIntent->toID().get() : nil);
```
This is probably not safe use of memory.  It creates a temporary RetainPtr<id>, destroys the strong reference, and keeps a raw pointer.  We want auto parent = m_parentIntent ? m_parentIntent->toID() : nullptr;

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.mm:
```diff
@@ -0,0 +1,126 @@
+    }
+}
+
+RetainPtr<id> CoreIPCPresentationIntent::toID() const
```
Can this be RetainPtr<NSPresentationIntent> instead of id?

mscottapple replied:
Honestly, I don't know. I was just following what the other CoreIPC cocoa implementations did. But since `toID()` is using this same declaration everywhere, I wouldn't think so.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.mm:
```diff
@@ -0,0 +1,126 @@
+    NSMutableArray<NSNumber *> *columnAlignments = nil;
+    switch (m_intentKind) {
+    case NSPresentationIntentKindParagraph:
+        result = [NSPresentationIntent paragraphIntentWithIdentity:m_identity nestedInsideIntent:parent];
```
If each of these changes to "return [NSPresentationIntent ...];" then we don't need a local variable result.

mscottapple replied:
Yes, you're right (except for the `NSPresentationIntentKindTable` case). This was an artifact of an earlier pass at this.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.mm:
```diff
@@ -0,0 +1,126 @@
+{
+    id parent = (m_parentIntent ? m_parentIntent->toID().get() : nil);
+    NSPresentationIntent *result = nil;
+    NSMutableArray<NSNumber *> *columnAlignments = nil;
```
This can be declared in the scope of case NSPresentationIntentKindTable like this:
case NSPresentationIntentKindTable: {
    auto columnAlignments = adoptNS([[NSMutableArray alloc] initWithCapacity:m_columnAlignments.size()]);
    ...
}

mscottapple replied:
Yes, it can, it just is ugly, IMHO.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.mm:
```diff
@@ -0,0 +1,126 @@
+        result = [NSPresentationIntent thematicBreakIntentWithIdentity:m_identity nestedInsideIntent:parent];
+        break;
+    case NSPresentationIntentKindTable:
+        columnAlignments = [[NSMutableArray alloc] initWithCapacity:m_columnAlignments.size()];
```
This is a memory leak.

mscottapple replied:
It's not, because of the `[columnAlignments release]` that was further down, but I've changed to use smart pointers to be consistent with The WK Way™.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.h:
```diff
@@ -0,0 +1,77 @@
+
+    int64_t m_intentKind { 0 };
+    int64_t m_identity { 0 };
+    std::unique_ptr<CoreIPCPresentationIntent> m_parentIntent { nullptr };
```
std::unique_ptr has a default constructor that initializes the pointer, so { nullptr } is unnecessary.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.h:
```diff
@@ -0,0 +1,77 @@
+
+#if PLATFORM(COCOA)
+
+#include "ArgumentCodersCocoa.h"
```
This also seems unnecessary.  <wtf/ArgumentCoder.h> should be sufficient.

mscottapple committed:
Generate serialization for NSPresentationIntent
https://bugs.webkit.org/show_bug.cgi?id=268779
rdar://109121874

Reviewed by achristensen07 (Alex Christensen).

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::typeFromObject):
(IPC::shouldEnableStrictMode):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.h: Added.
(WebKit::CoreIPCPresentationIntent::CoreIPCPresentationIntent):
* Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.mm: Added.
(WebKit::CoreIPCPresentationIntent::CoreIPCPresentationIntent):
(WebKit::CoreIPCPresentationIntent::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPresentationIntent.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCTypes.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST):

Canonical link: https://commits.webkit.org/274269@main
